enhancements/run-bundle.md

```diff
@@ -0,0 +1,263 @@
+   Can we change that to using the discovery API?
+   * Based on discussions, using the discovery API is the way to go.
+
+2. How do we handle --mode= flag on registry add
+   * should we ALWAYS default to --mode=semver?
+     * this has less testing and isnot yet the default. We should stick with
+        the default.
+   * should we have a flag that allows our default mode handling to be overridden?
+     * no
+
+## Summary
+
+Enhance the Operator SDK `run` command to support the new bundle format.
+
+## Motivation
+
+One of the primary goals of the Operator SDK is to make it easy for operator
+developers to run and test their operators with Operator Lifecycle
+Manager (OLM). There currently exists a `run --olm` command which supports the
+packagemanifest format. In order to add support for the new bundle format, we
+are migrating to subcommands on the run command instead of increasing the
+number of flags.
+
+### Goals
+
+Provide a command line that works with bundle image format.
+
+### Non-Goals
+
+This design does not cover the replacement of run --olm with run
+packagemanifest. The change to `run packagemanifest` is covered by
+[OSDK-1126](https://issues.redhat.com/browse/OSDK-1126).
+
+## Proposal
+
+### Implementation Details/Notes/Constraints [optional]
+
+The Operator SDK currently supports running an operator with OLM using the
+`run` command with the `--olm` flag. The existing command only supports the
+package manifest format. This enhancement will add a new subcommand, `bundle`,
+to the `run` command. This new command, `run bundle`, will support the new
+bundle format. More information about the different formats can be found at
+the URLs below:
+
+ * Operator bundles: https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md
+ * Package manifests format: https://github.com/operator-framework/operator-registry/tree/v1.5.3/#manifest-format
+
+The idea is to consolidate the run command into different subcommands to make
+things easier to understand and eliminate flag bloat.
+
+ * `run local` - exists today
+ * `run packagemanifest` - covered by [OSDK-1126](https://issues.redhat.com/browse/OSDK-1126)
+ * `run bundle` - covered by this design document
+
+#### Command Line Interface
+The new command line will look as follows:
+
+```
+operator-sdk run bundle <bundle-image> [--index-image=] [--namespace=] [--install-mode=(AllNamespace|OwnNamespace|SingleNamespace=)]
+```
+
+* `<bundle-image>` is a positional argument that specifies the bundle image.
+* `[--index-image=]` is an optional flag that specifies the index image.
+  * If `--index-image` is not provided, default to
+    `quay.io/operator-framework/upstream-opm-builder`
+  * If `--index-image` is provided, use the provided value
+* `[--namespace=]` is an optional flag that specifies the namespace to install
+  the operator. It will default to the `KUBECONFIG` context.
+* `[--install-mode=]` is an optional flag that specifies the
```

Comment:
The `installModes` are already defined by the bundle. I'm assuming this flag is for determining what OperatorGroup to generate - in that case, all we need is a `--target-namespaces=[]` arg taking the set of namespaces to watch (which matches the terminology in the OperatorGroup)

Reply:
We want someone to select the install mode to test not interface with the operator group I think was the the thinking

Reply:
My point is that flag may be confusing. The flag is called `installMode` and the values are `installMode` values, but it does not set the `installMode` on the operator (it can't, because that is already set in the bundle, and this command doesn't build the bundle).

Reply:
A user can set `--install-mode=InstallModeType[=nsY]`. If a bundle's CSV supports that mode in `installModes` by comparing type strings, `run bundle` will either create an OperatorGroup in some namespace targeting `nsY`, or use an existing and compatible OperatorGroup.

I see where this can create confusion: a supported mode in `installModes` is only a statement about runtime availability, whereas `--install-mode` mixes in OperatorGroup semantics.

I don't think it makes sense to pollute the CLI with two flags, one for `installMode` and another for namespaces to target/watch. We definitely need the latter, and we want the former so users see a mapping between the way they're testing operator deployment and what their CSV/bundle encodes, even though we could infer mode from passed namespaces.